hspec-core/src/Test/Hspec/Core/Spec.hs

```diff
@@ -44,8 +51,23 @@ describe label spec = runIO (runSpecM spec) >>= fromSpecList . return . specGrou
 -- > describe "absolute" $ do
 -- >   it "returns a positive number when given a negative number" $
 -- >     absolute (-1) == 1
-it :: Example a => String -> a -> SpecWith (Arg a)
-it label action = fromSpecList [specItem label action]
+it :: (?loc :: CallStack, Example a) => String -> a -> SpecWith (Arg a)
+it label action = do
+  cwd <- runIO getCurrentDirectory
+  addLocation cwd $ fromSpecList [specItem label action]
+  where
+    addLocation cwd = mapSpecItem_ (\item -> item { itemLocation = toLocation cwd <$> srcLoc })
+
+    -- The IP type signatures on ?loc here and on ‘it’ itself as well
+    -- as the actual IP names are *essential*: if the names differ or
+    -- one of the signatures is missing, IP scoping rules or CallStack
+    -- rules will mess the stack up and we won't get what we wanted.
+    srcLoc = case reverse (getCallStack (?loc :: (?loc :: CallStack) => CallStack)) of
+      ("it", loc):_ -> Just loc
+      _ -> Nothing
+
+    toLocation cwd s =
+      Location (makeRelative cwd $ srcLocFile s) (srcLocStartLine s) (srcLocStartCol s) ExactLocation
```

Comment:
How exactly does the path look if you don't apply `makeRelative` here?  Note that the `cwd` may be different from the source code location at the time the test suite is run.  If we can decide on the final path statically rather than doing it at run time, I would prefer that.